power_grid_model_c/power_grid_model/include/power_grid_model/sparse_mapping.hpp

```diff
@@ -54,31 +54,92 @@ inline SparseMapping build_sparse_mapping(IdxVector const& idx_B_in_A, Idx const
     auto const n_A = static_cast<Idx>(idx_B_in_A.size());
     using SparseEntry = std::pair<Idx, Idx>;
     std::vector<SparseEntry> entries(n_A);
+
     std::transform(idx_B_in_A.cbegin(), idx_B_in_A.cend(), IdxCount{0}, entries.begin(), [](Idx j_B, Idx i_A) {
         return SparseEntry{i_A, j_B};
     });
-    // initialize result
+
     SparseMapping sparse_mapping;
     sparse_mapping.indptr.resize(n_B + 1);
     sparse_mapping.reorder.resize(n_A);
-    // counting sort column, keep row as sorted
+
     IdxVector counter(n_B, 0);
     for (auto const& entry : entries) {
         ++counter[entry.second];
     }
-    // accumulate value to indptr
+
     std::inclusive_scan(counter.cbegin(), counter.cend(), sparse_mapping.indptr.begin() + 1);
-    // copy back
+
     std::copy(sparse_mapping.indptr.cbegin() + 1, sparse_mapping.indptr.cend(), counter.begin());
-    // sort
+
     for (auto it_entry = entries.crbegin(); it_entry != entries.crend(); ++it_entry) {
         sparse_mapping.reorder[--counter[it_entry->second]] = it_entry->first;
     }
+
     assert(sparse_mapping.indptr[0] == 0);
     assert(sparse_mapping.indptr.back() == n_A);
     return sparse_mapping;
 }
 
-} // namespace power_grid_model
+struct DenseMapping {
+    IdxVector indvector;
+    IdxVector reorder;
+};
+
+inline DenseMapping build_dense_mapping(IdxVector const& idx_B_in_A, Idx const n_B) {
+    auto const n_A = static_cast<Idx>(idx_B_in_A.size());
+
+    using DenseEntry = std::pair<Idx, Idx>;
+    DenseMapping dense_mapping;
+
+    if (static_cast<double>(n_A) * log(static_cast<double>(n_A)) <=
+        static_cast<double>(n_A) + static_cast<double>(n_B)) {
+
+        std::vector<DenseEntry> mapping_to_from;
+
+        std::transform(idx_B_in_A.begin(), idx_B_in_A.end(), IdxCount{0}, std::back_inserter(mapping_to_from),
+                       [](Idx value, Idx orig_idx) {
+                           return std::pair{value, orig_idx};
+                       });
 
+        std::ranges::sort(mapping_to_from);
+
+        std::ranges::transform(mapping_to_from, std::back_inserter(dense_mapping.indvector),
+                               [](DenseEntry const& to_from) { return to_from.first; });
+
+        std::ranges::transform(mapping_to_from, std::back_inserter(dense_mapping.reorder),
+                               [](DenseEntry const& to_from) { return to_from.second; });
+    } else {
+        std::vector<DenseEntry> entries(n_A);
+
+        std::transform(idx_B_in_A.cbegin(), idx_B_in_A.cend(), IdxCount{0}, entries.begin(), [](Idx j_B, Idx i_A) {
```

Comment:
i believe this could also be

```suggestion
        std::ranges::transform(idx_B_in_A, IdxCount{0}, entries.begin(), [](Idx j_B, Idx i_A) {
```

Reply:
this is not possible due to `boost::counting_iterator` not satisfying `std::forward_iterator`. We can't switch to `std::views::iota` due to `boost::iterators::zip_iterator` not supporting the correct implementation. deadlock much.